// structure/src/add-thumbnail-path.js
var settings = require('settings');

module.exports = function(structure){
	var thumbnailsPath = settings.thumbnails || 'media/images/common/thumbs/';

	Object.keys(structure.slides).forEach(function(slideId){
		var slide = structure.slides[slideId],
			location = isAbsolutePath(slide.template) ? getLocation(slide.template) : '';

		slide.thumbnail = location + settings.media + thumbnailsPath + getScenarioPath(slide.thumbnail) + getThumbnailName(slide) + '.jpg'
	});

	return structure;
};

function getScenarioPath(thumbnailPath){
	if(settings.isScenario){
		return thumbnailPath || '';
	}

	return '';
}

function getThumbnailName(slide){
	return isBuilt(slide) ? getProdSlideId(slide) : getDevSlideId(slide);
}

function getDevSlideId(slide){
	var path = slide.template;

	return path.substring(path.lastIndexOf("/") + 1, path.lastIndexOf("."));
}

function getProdSlideId(slide){
	var path = slide.template.split('/');

	return path[path.length - 2];
}

function isBuilt(slide){
	return /slides\/[^\/]+\/index.html/.test(slide.template);
}

function isAbsolutePath(path){
	return /^((file|http|https):\/\/)|^\//.test(path);
}

function getLocation(file){
	return file.substring(file.lastIndexOf("/") + 1, 0);
}
