// core/dist/js/cobalt-api/src/ViewObservationLifeCycle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _componetsDOMObserver = require('./componetsDOMObserver.js');

var _componetsDOMObserver2 = _interopRequireDefault(_componetsDOMObserver);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var ViewObservationLifeCycle = function () {
	function ViewObservationLifeCycle(dispatcher) {
		var _this = this;

		_classCallCheck(this, ViewObservationLifeCycle);

		this.subscriber = null;

		dispatcher.subscribeAction(function (action) {
			switch (action.type) {
				case dispatcher.ActionType.EDIT_PRESENTATION:
					if (_this.subscriber) {
						_this.reset();
					}
					_this.init(dispatcher);
					break;
				case dispatcher.ActionType.VIEW_PRESENTATION:
					_this.reset();
					break;
			}
		});
	}

	_createClass(ViewObservationLifeCycle, [{
		key: 'init',
		value: function init(dispatcher) {
			this.subscriber = _componetsDOMObserver2.default.subscribe(function (component) {
				dispatcher.dispatchEvent({
					type: dispatcher.EventType.DOM_REFRESHED,
					element: component
				});
			});
		}
	}, {
		key: 'reset',
		value: function reset() {
			this.subscriber.unsubscribe();
			this.subscriber = null;
		}
	}]);

	return ViewObservationLifeCycle;
}();

exports.default = ViewObservationLifeCycle;
