// nav/index.js
var spa = require('spa'),
	settings = require('settings'),
	utils = require('utils'),
	isNextLocked, isPrevLocked;

function spaSwipeBlocker(event){
	event.stopPropagation();
}

if(spa.goto){
	exports.onenter = function(callback){
		document.addEventListener('slideenter', callback);
	};

	exports.onleave = function(callback){
		document.addEventListener('slideleave', callback);
	};

	exports.goto = function(options){
		exports.observer.publish("willGoto", options);
		spa.goto(options);
	};

	exports.nextSlide = function(){
		spa.nextSlide();
	};

	exports.previousSlide = function(){
		spa.previousSlide();
	};

	exports.openPopup = function(opts){
		spa.openPopup(opts);
	};

	exports.closePopup = function(){
		spa.closePopup();
	};

	exports.collectPopup = function(slides, slide){
		spa.collectPopup(slides, slide);
	};

	exports.lockNext = function(){
		if(!isNextLocked){
			document.addEventListener('swipeleft', spaSwipeBlocker, true);
			isNextLocked = true;
		}
	};

	exports.lockPrev = function(){
		if(!isPrevLocked){
			document.addEventListener('swiperight', spaSwipeBlocker, true);
			isPrevLocked = true;
		}
	};

	exports.unlockNext = function(){
		document.removeEventListener('swipeleft', spaSwipeBlocker, true);
		isNextLocked = false;
	};

	exports.unlockPrev = function(){
		document.removeEventListener('swiperight', spaSwipeBlocker, true);
		isPrevLocked = false;
	};


	exports.lock = function(callback){
		exports.lockNext();
		exports.lockPrev();
	};

	exports.unlock = function(callback){
		exports.unlockNext();
		exports.unlockPrev();
	};

	exports.close = function(){
		spa.viewport.dispatchSlideLeaveForCurrentSlide();
		parent.location.href = "components/nav/src/closedialog.html";
	};
}else{
	exports.onenter = function(callback){
		setTimeout(callback, 100);
	};

	exports.onleave = function(callback){
		window.addEventListener('unload', callback);
	};

	exports.goto = function(options){
		if(options.slide){
			exports.observer.publish("willGoto", options);
			location.href = options.slide + '.html';
		}
	};
}

// Do not delete 'top'. This is a fix to open pdf on eWizard-Mobile
exports.openDocument = function(path){
	if(window.top !== window){
		window.top.open(path, "_blank");
	}else{
		window.open(path, "_blank");
	}
};

exports.sendMail = function(options){
	var utils = require('utils'),
		mailString = 'mailto:{to}?subject={subject}&body={body}&attachment={attachment}';
	mailString = utils.template(mailString, options);
	window.open(mailString, '_blank');
};

exports.observer = new utils.Observer();
