// core/src/js/cobalt/index.js
import rivets from 'rivets'
import { declareDefaultBindings, declareImportsBindings } from 'extension'
import { setElementData } from 'model'
import utils from 'utils'
import params from 'params'
import { CommonModel } from 'general-model'

export default class Cobalt {
  constructor (scope, localization) {
    this.scope = scope

    this.configure()
    this.extendRivets(scope)
    setElementData(document, scope.m)
  }

  configure() {
    rivets.configure({ params })
  }

  extendRivets (scope) {
    declareDefaultBindings(rivets, scope)
    declareImportsBindings(rivets, scope)
  }

  initialize (rootElement) {
    delete rivets.components['rv-template']
    this.mainView = this.bind(rootElement)
    this.initCommonModel()

    var onWillgoto = function() {
      this.mainView.unbind();
      this.mainView.bindings = [];
      this.mainView.els = [];
      this.mainView.binders = {};
      this.mainView.components = {};
    }.bind(this);

    document.addEventListener('willgoto', onWillgoto);

    return Promise.resolve()
  }

  initCommonModel() {
    new CommonModel(this.scope.m.common);
  }

  bind (element, scope = this.scope) {
    return rivets.bind(element, scope)
  }
}
