// co-select/src/js/utils.js
import * as utils from 'utils';

module.exports.observe = function (element, config, callback) {
	new MutationObserver(function (mutations) {
		callback(mutations);
	}).observe(element, config);
	callback([]);
};

module.exports.getOptions = function (element) {
	return utils.toArray(
					element.getElementsByTagName('co-option'))
					.map((optionElement) => optionElement.model)
	.filter((optionElementModel) => optionElementModel);
};

module.exports.isThisComponentInCheckbox = function (element) {
	return element.parentElement.tagName == "CO-CHECKBOX-GROUP" || element.parentElement.tagName == "CO-CHECKBOXES";
};

module.exports.getCoOptionFromCollectionItemInMutations = function (mutation) {
	if (mutation.length && mutation[0].tagName == 'CO-COLLECTION-ITEM') {
		return mutation[0].querySelector('co-option');
	}
	return false;
};

function isComponentLoad(component) {
	return component.hasAttribute('co-component');
}
module.exports.isCollectionLoad = function (collection) {
	if (collection && collection.children.length > 0) {
		return isComponentLoad(collection.children[0]);
	} else {
		return true;
	}
};
