// prefixer/index.js
function getSupportedProperty(featureName){
	var domPrefixes = 'Webkit Moz ms O'.split(' '),
		element = document.createElement('div'),
		featurenameCapital;

	if(element.style[featureName] !== undefined){
		return featureName;
	}else{
		featurenameCapital = featureName.charAt(0).toUpperCase() + featureName.substr(1);
		for(var i = 0; i < domPrefixes.length; i++){
			if(element.style[domPrefixes[i] + featurenameCapital] !== undefined){
				return domPrefixes[i] + featurenameCapital;
			}
		}
	}
	return featureName;
}

function getSupportedEvent(map){
	var element = document.createElement('div');

	for(var key in map){
		if(map.hasOwnProperty(key) && element.style[key] !== undefined){
			return map[key];
		}
	}
}

exports.getSupportedProperty = getSupportedProperty;

exports.setStyle = function(element, property, value){
	element.style[getSupportedProperty(property)] = value;
};

exports.events = {
	transitionend: getSupportedEvent({
		'transition'       : 'transitionend',
		'WebkitTransition' : 'webkitTransitionEnd'
	}),
	animationend: getSupportedEvent({
		'animation'       : 'animationend',
		'WebkitAnimation' : 'webkitAnimationEnd'
	}),
	animationstart: getSupportedEvent({
		'animation'       : 'animationstart',
		'WebkitAnimation' : 'webkitAnimationStart'
	})
};
