// spa/src/utils.js
exports.getFirstEnabledSlide = function (structure, chapter) {
  return getChapterContent(structure, chapter)
    .filter(isEnableSlide)[0]
}

exports.getFirstNotEmptyChapter = function (structure, chapter) {
  var startIndex = chapter ? structure.storyboard.indexOf(chapter) : 0;
  return structure.storyboard.slice(startIndex)
    .filter(isEnableChapter)
    .filter(isEnableSlidesInChapter(structure))[0]
}

exports.isEnableSlide = isEnableSlide;

function isEnableSlidesInChapter(structure, chapter){
  return function (chapter){
    return getChapterContent(structure, chapter).some(isEnableSlide)
  }
}

function getChapterContent(structure, chapter) {
  try {
    return structure.chapters[chapter].content
  } catch(e) {
    throw new Error('The chapter ' + chapter + ' is absent in the structure.');
  }
}

function isEnableChapter (chapter){
  return isEnableStructureEntity(chapter)
}

function isEnableSlide(slide){
  return isEnableStructureEntity(slide)
}

function isEnableStructureEntity (entity){
  var DISABLE_LABEL = "!";
  return !!entity.indexOf(DISABLE_LABEL)
}
